From 18f4e35d55edcb2a53fecfa594a74630e8859eb3 Mon Sep 17 00:00:00 2001
From: Monika Korada <quic_koramoni@quicinc.com>
Date: Thu, 21 Mar 2024 15:43:45 +0530
Subject: [PATCH] [WAR] hostapd: fix uninitialized pointer, allowed_freqs
access

The pointer, allowed_freqs was uninitialized and
dereferenced from int_array_add_unique API. So it
is set to NULL before the API is called and freed
only if it is valid.

Also, this API is needed only for QCA CLD driver.
ath platform won't use this. Hence protected this
API call with CONFIG_DRIVER_NL80211_QCA flag.

Signed-off-by: Monika Korada <quic_koramoni@quicinc.com>
---
 src/ap/beacon.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2725,8 +2725,11 @@ static int __ieee802_11_set_beacon(struc
 	struct hostapd_config *iconf = iface->conf;
 	struct hostapd_hw_modes *cmode = iface->current_mode;
 	struct wpabuf *beacon, *proberesp, *assocresp;
-	int res, ret = -1, i;
+	int res, ret = -1;
+#ifdef CONFIG_DRIVER_NL80211_QCA
+	int i;
 	struct hostapd_hw_modes *mode;
+#endif
 
 	if (!hapd->drv_priv) {
 		wpa_printf(MSG_ERROR, "Interface is disabled");
@@ -2805,6 +2808,8 @@ static int __ieee802_11_set_beacon(struc
 				    iconf->center_freq_device) == 0)
 		params.freq = &freq;
 
+	params.allowed_freqs = NULL;
+#ifdef CONFIG_DRIVER_NL80211_QCA
 	for (i = 0; i < hapd->iface->num_hw_features; i++) {
 		mode = &hapd->iface->hw_features[i];
 
@@ -2817,6 +2822,7 @@ static int __ieee802_11_set_beacon(struc
 						   iconf->acs_ch_list.num),
 						 true, &params.allowed_freqs);
 	}
+#endif /* CONFIG_DRIVER_NL80211_QCA */
 
 	res = hostapd_drv_set_ap(hapd, &params);
 	hostapd_free_ap_extra_ies(hapd, beacon, proberesp, assocresp);
